Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -102,14 +121,33 @@ private async Task<ReadHttpResponseMessage> SendAsync<TRequest>(RestMethod restM
             }
         }
 
-        private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod method)
+        private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod restMethod, CancellationToken combinedCancellationToken)
         {
             Uri hostUri = request.RequestUri.IsAbsoluteUri ? request.RequestUri : _httpClient.BaseAddress;
             string schemeAndAuthority = hostUri.GetLeftPart(UriPartial.Authority);
 
             var metadata = new Metadata();
-            var context = new AuthInterceptorContext(schemeAndAuthority, method.FullName);
-            await _channelAuthInterceptor(context, metadata).ConfigureAwait(false);
+            var context = new AuthInterceptorContext(schemeAndAuthority, restMethod.FullName);
+
+            var combinedCancellationTask = Task.Delay(-1, combinedCancellationToken);
+            var channelTask = _channelAuthInterceptor(context, metadata);
+            var resultTask = await Task.WhenAny(channelTask, combinedCancellationTask).ConfigureAwait(false);
+            if (combinedCancellationTask == resultTask)
```

Comment:
I think that if this is the case, then `resultTask` will be faulted and `resultTask.Exception` will contain a `TaskCanceledException`. The only possible state in which combinedCancellationTask may run to completion is faulted.

Or am I missing something?

My point is that I think lines from 135 to 149 may be substituted by:

```csharp
if(resultTask.IsFaulted)
{
    ExceptionDispatchInfo.Capture(resultTask.Exception).Throw();
}
```
Or what am I missing?

Reply:
No, I think you're right. (I thought before that we'd want to provide a different exception on timeout for auth headers, but in this case I think it's fine for the task cancellation to just propagate.)

Reply:
The `Task.Delay(-1, combinedCancellationToken)` does not get faulted and does not have an exception when the cancellation gets called on the token. It gets the status: Cancelled and that's it.

Reply:
Ah, then you can change lines from 135 to 149 with:

```
await resultTask;
```

This will throw the TaskCancelledException if cancellation happened, or whatever exception happened if the real task failed.
If you get past this line of code, then there was no error or cancellation.

I think that will be cleaner and easier to understand than all the code we have now.

Reply:
Yes, this makes sense. Updated.